doc/kj.md

```diff
@@ -0,0 +1,84 @@
+---
+layout:	page
+title:	KJ
+---
+
+# KJ: A Performance and Safety Library for C++
+
+KJ provides a variety of types, templates, and macros to ease the writing of correct,
+high performance C++ code.
+
+## Primitives
+
+### Maybe (optional)
+
+`Maybe<T>` is a template for an optionally contained value, similar to (but predating)
+[`std::optional<T>`](https://en.cppreference.com/w/cpp/utility/optional). It has the advantage of
+ensuring that the empty case is handled, thus preventing null pointer dereferences and/or access to
+invalid objects at runtime.
+
+You can construct a `Maybe<T>` using either a value of type `T` or using `nullptr`:
+
+```c++
+#include <kj/common.h>
+
+auto number = kj::Maybe<double>(1.0);
+auto empty = kj::Maybe<double>(nullptr);
+```
+
+Afterwards, you can simultaneously check for and access the contents of a `Maybe<T>` using the
+`KJ_IF_MAYBE()` macro as follows:
+
+```c++
+KJ_IF_MAYBE(pointer_to_value, maybe_value) {
+  doSomething(*pointer_to_value);
+} else {
+  handleEmptyCase();
+}
+```
+
+### OneOf (variant)
+
+`OneOf<T>` is a template for a type-safe tagged union, similar to (but predating)
+[`std::variant`](https://en.cppreference.com/w/cpp/utility/variant). It has the advantage of
+enabling easy handing of the different control flow paths with compiler warnings automatically for
+unhandled cases using a familiar `switch`/`case` approach.
+
+You can construct a `OneOf<T>` using a value valid for any of the types:
+
+```c++
+#include <kj/one-of.h>
+
+auto numberInt = kj::OneOf<int, double, kj::String, bignum>(8);
+auto numberFloat = kj::OneOf<int, double, kj::String, bignum>(8.0);
+auto numberString = kj::OneOf<int, double, kj::String, bignum>("8");
```

Comment:
Same deal as with the `Maybe`, it's more natural in KJ (I think) to define a `OneOf` type that you later use, which might be useful to show here rather than `auto`.

Reply:
WRT this and [the above](https://github.com/capnproto/capnproto/pull/1141#discussion_r549881686), I think my main intent there is that reading-wise I always these kinds of one-liners to be the most succinctly informative.

I always like more examples though, so I'll add class/struct examples too that illustrate what you're talking about.